867/Src/quadlog-gampsi/man/dran.man
.TH DRAN 3 "14 July 2000" "Version 1.00"
.\" WARNING: This file was produced automatically from file common/dran.f
.\" by fortran-to-man-page.awk on Sun Dec 31 09:02:09 MST 2000.
.\" Any manual changes will be lost if this file is regenerated!
.SH NAME
dran \- Double-precision pseudo-random number
.\"=====================================================================
.SH SYNOPSIS
Fortran (77, 90, 95, HPF):
.RS
.B f77
.I "[ flags ] file(s) .\|.\|. -L/usr/local/lib -lgjl"
.RS
.nf
.B "DOUBLE PRECISION FUNCTION dran()"
.fi
.RE
.RE
C (K&R, 89, 99), C++ (98):
.RS
.B cc
.I "[ flags ] -I/usr/local/include file(s) .\|.\|. -L/usr/local/lib -lgjl"
.br
Use
.RS
.B "#include <gampsi.h>"
.RE
to get this prototype:
.RS
.B "fortran_double_precision dran(void);"
.RE
.RE
.PP
NB: The definition of C/C++ data types
.B fortran_
.IR xxx ,
and the mapping of Fortran external names to C/C++ external names,
is handled by the C/C++ header file.  That way, the same function
or subroutine name can be used in C, C++, and Fortran code,
independent of compiler conventions for mangling of external
names in these programming languages.
.PP
Last code modification: 14-Jul-2000
.\"=====================================================================
.SH DESCRIPTION
Generate and return a double-precision pseudo-random number from
the interval (0.0, 1.0). The significand of the returned value
should have about 58 pseudo-random bits. This is sufficient for
most arithmetic systems: in IEEE 754 arithmetic, the fractional
part of the double-precision significand has only 52 bits.
.PP
The initial generator seed is the same on the first call to this
function after every program startup, so that the sequence of
pseudo-random number is reproducible. This routine has no
provision for the user to alter the initial seed.
.PP
The algorithm is based on ``ACM Algorithm 266: Pseudo-Random
Numbers'', by M. C. Pike and I. D. Hill, Communications of the
ACM, Vol. 8, No. 10, 605--606, October 1965, modified by
Hansson, and later used in the book ``Software Manual for the
Elementary Functions'', by W. J. Cody, Jr. and W. Waite,
Prentice-Hall (1980), ISBN 0-13-822064-6.
.\"=====================================================================
.SH "SEE ALSO"
.BR ran (3).
.\"=====================================================================
.SH AUTHORS
The algorithms and code are described in detail in
the paper
.RS
.I "Algorithm xxx: Quadruple-Precision Gamma(x) and psi(x) Functions for Real Arguments"
.RE
in ACM Transactions on Mathematical Software,
Volume ??, Number ??, Pages ????--???? and
????--????, 2001, by
.RS
.nf
Nelson H. F. Beebe
Center for Scientific Computing
University of Utah
Department of Mathematics, 110 LCB
155 S 1400 E RM 233
Salt Lake City, UT 84112-0090
Tel: +1 801 581 5254
FAX: +1 801 581 4148
Email: \fCbeebe@math.utah.edu\fP, \fCbeebe@acm.org\fP, \fCbeebe@computer.org\fP
WWW URL: \fChttp://www.math.utah.edu/~beebe\fP
.fi
.RE
and
.RS
.nf
James S. Ball
University of Utah
Department of Physics
Salt Lake City, UT 84112-0830
USA
Tel: +1 801 581 8397
FAX: +1 801 581 6256
Email: \fCball@physics.utah.edu\fP
WWW URL: \fChttp://www.physics.utah.edu/people/faculty/ball.html\fP
.fi
.RE
.\"==============================[The End]==============================
.\"=====================================================================
.\" This is for GNU Emacs file-specific customization:
.\" Local Variables:
.\" fill-column: 50
.\" End:


Changes
April 22 2021
   * Added 1013-1014 from Volume 47, number 1, (January 2021)
   * Added 1012 from Volume 46, number 4, (November 2020)
October 19 2020
   * Added 1011 from Volume 46, number 3, (September 2020)
August 7 2020
   * Added 1003-1006 from Volume 46, number 1, (April 2020)
   * Added 1007-1010 from Volume 46, number 2, (June 2020)
January 12 2020
   * Updated 999 at the request of the author -- updated user manual supplied.
January 7 2020
   * Added 999-1002 from Volume 45, number 4, (December 2019)
October 13 2019
   * Added 995-998 from Volume 45, number 3, (August 2019)
August 8 2019
   * Added 993 and 994 from Volume 45, number 2, (June 2019)
August 7 2019
   * Added 991 and 992 from Volume 45, number 1, (February 2019)
February 1 2019
   * Added 989 and 990 from Volume 44, number 3, (August 2018)
May 8 2018
   * Added 987 and 988 from Volume 44, number 3, (April 2018)
October 13 2017
   * Added 986 from Volume 44, number 2, (October 2017)
   * Corrected publication month for Algs 979-985 (now Oct).
September 30 2017
   * Added 979-985 from Volume 44, number 2, (September 2017)
   * Added 975-978 from Volume 44, number 1, (July 2017)
April 21 2017
   * Added 973, 974 and Remark on 936 from Volume 43, number 4 
     (March 2017)
February 3 2017
   * Added 968-972 from Volume 43, number 3 (January 2017)
September 16 2016
   * Added 965 from Volume 43, number 1 (August 2016)
   * Added 966, 967 from Volume 43, number 2 (September 2016)
   * Removed files 620 and 620.bib (all other algorithm files
     are only available as compressed files.
   * Replaced index.gz and index.html.gz by .zip files
September 14 2016:
   * Added 100.zip and 101.zip. These files are identical as
   the two algorithms are inter-related and the driver program
   uses both codes.
August 30 2016:
   * Added 963-964 from Volume 42, number 4 (July 2016)
July 23 2016
   * Updated contact information on CALGO web pages.
June 19 2016:
   * Added 960-962 from Volume 42, number 3 (June 2016)
   * Updated 916 to include the code submitted with Remark 
     in Volume 42, number 3 (Jun 2016)
June 15 2016:
   * Added 957 from Volume 42, number 1 (February 2016)
   * Added 958-959 from Volume 42, number 2 (June 2016)
May 30 2016
   * Added 955-956 from Volume 42, number 1 (February 2016)
May 27 2016:
   * Added new code supplied as Remark to Algorithm 673
     appearing in Volume 42, number 1.
December 22 2015:
   * Added 947-954 from Volume 41, numbers (1-4).
August 29 2015:
   * Mended a number of odd/broken compressed files that were on
     the CALGO site:
       755.gz, 756.gz, 823.gz replaced by zip files
       709.gz, 709a.gz, 709b.gz replaced by 709.zip
       573.gz, 573d.gz, 573s.gz replaced by 573.zip
       358.gz, 805.gz, 830.gz, 832.zip all updated
       620.zip added -- this contains all the 620 components.
September 26 2014
   * Updated 897 following accepted remark.
September 1 2014
   * Added 943-946 from Volume 40, number 4 (June 2014)
May 17 2014
   * Fixed problem with Algorithm 917 missing from indexes.
May 5 2014
   * Added 935-938 from Volume 40, number 2 (February 2014)
   * Added 939-942 from Volume 40, number 3 (April 2014)
October 9 2013
   * Added 929, 930 from volume 39, number 4 (July 2013)
   * Added 931-934 from volume 40, number 1 (October 2013)
July 18 2013
   * Added 928 from volume 39, number 3 (May 2013)
   * Updated 905 from code supplied by author thackerw@winthrop.edu
March 3 2013
   * Added 926-927 from volume 39, number 2 (February 2013)
January 25 2013
   * Replaced 923 after editing Licence file -- author request
January 3 2013
   * Added 924-925 from volume 39, number 1 (November 2012)
September 10 2012
   * Added 920-923 from volume 38, number 4 (August 2012)
June 17 2012
   * Added 917-919 from volume 38, number 3 (April 2012)
March 14 2012
   * Corrected typo in index files (Alg 916)
February 29 2012
   * Replaced 916 -- old version of code uploaded in error.
February  5 2012
   * Added 916 from Volume 38, number 2 (December 2012)
December 30 2011
   * Replaced 905 due to corrupt blas.f file in the original (pointed
     out by Ronald.M.Bloom@aero.org).
December 7 2011
   * Added 913-915 from Volume 38, number 1 (November 2011)
   * Updates 778 as per Remark in Volume 38, No. 1
April 12 2011
   * Added 909-911 from volume 37, number 4 (April 2011)
October 24 2011
   * Changed Algorithm 419 in line with changes from Remark on Algorithm 419 by
     David H. Withers CACM (March 1974) Vol 17 No 3 p. 157
     Pointed out by j.a.horsmeier@wanadoo.nl (Jos)
October 11 2010
   * Added 904-908 from volume 37, number 3 (September 2010)
May  20 2010
   * Added 902, 903 from volume 37, number 2 (April 2010)
Feb  27 2010
   * Added 898, 899, 900, 901 from volume 37, number 1 (January 2010)
July 29 2009
   * Added 895, 896, 897 from volume 36, number 3 (July 2009)
Apr  6 2009
   * Added 894 from volume 36, number 2 (March 2009)
Mar  30 2009
   * Added 891, 892, 893 from volume 36, number 1 (March 2009)
Feb  12 2009
   * Added 890 from volume 35, number 4 (February 2009)
Jan   6 2009
   * Added 885-889 from volume 35, number 3 (October 2008)
Sept  4 2008
   * Added 881-884 from volume 35, number 2 (July 2008)
July 25 2008
   * Added 876-877 from volume 34, number 4 (July 2008)
   * Added 878-880 from volume 35, number 1 (July 2008)
June 11 2008
   * Added 873-875 from volume 34, number 2 and 3 (March and May 2008)
Feb   8 2008
   * Added 867-869 from volume 33, numbers 3 and 4 (August 2007)
   * Added 870-872 from volume 34, number 1 (January 2008)
Aug   3 2007
   * Added 863-866 from volume 33, numbers 1 and 2.
July  3 2007
   * Updated 620 to include GAMS classifications to algorithms 814-862 inc.
Jan  10 2007
   * Updated 644 as a result of an accepted remark by Kodama
Jan   5 2007
   * Added 857-862 from volume 32, 4 (December 2006)
Nov  21 2006
   * Added 358 -- donated by John Burkardt <burkardt@csit.fsu.edu>.
Oct   5 2006
   * Added 854 from volume 32, 2 (June 2006)

867/Src/quadlog-gampsi/common/dran.f
      DOUBLE PRECISION FUNCTION dran()
************************************************************************
*     (Double-precision pseudo-random number)
*     Generate and return a double-precision pseudo-random number from
*     the interval (0.0, 1.0).  The significand of the returned value
*     should have about 58 pseudo-random bits.  This is sufficient for
*     most arithmetic systems: in IEEE 754 arithmetic, the fractional
*     part of the double-precision significand has only 52 bits.
*
*     The initial generator seed is the same on the first call to this
*     function after every program startup, so that the sequence of
*     pseudo-random number is reproducible.  This routine has no
*     provision for the user to alter the initial seed.
*
*     The algorithm is based on ``ACM Algorithm 266: Pseudo-Random
*     Numbers'', by M. C. Pike and I. D.  Hill, Communications of the
*     ACM, Vol. 8, No. 10, 605--606, October 1965, modified by
*     Hansson, and later used in the book ``Software Manual for the
*     Elementary Functions'', by W. J. Cody, Jr. and W. Waite,
*     Prentice-Hall (1980), ISBN 0-13-822064-6.
*     [14-Jul-2000]
************************************************************************
*
*     The single precision version of this subprogram is intended for
*     use on computers with fixed-point wordlength of at least 29
*     bits.  It is best if the floating point significand has at most
*     29 bits.
*
*     Following Cody and Waite's recommendation (p. 14), we produce a
*     pair of random numbers and use ran1 + 2**(-29)*ran2 in an
*     attempt to generate about 58 random bits.
*     [14-Jul-2000]
************************************************************************
*
*     Intrinsic functions
*
      INTRINSIC           dble
*
*     Built-in functions
*
      DOUBLE PRECISION    dble
*
*     External functions
*
      EXTERNAL            dfloat,      dstorf
*
      DOUBLE PRECISION    dfloat,      dstorf
*
*     Local variables
*
      INTEGER             iy
*
*     The variable iy must be saved to provide a memory across calls
*     to this function:
*
      SAVE                iy
*
      DATA                iy     /100001/
*
*     iy is always in the range (1..2796202) <= (1..2^22), so that it
*     can be represented exactly when converted to a single-precision
*     floating-point number in DEC VAX or IEEE 754 arithmetic.  That
*     conversion might lose 1 bit in IBM S/360 arithmetic, due to its
*     wobbling precision (because of hexadecimal normalization) of 21
*     to 24 fraction bits.
*
      iy = iy * 125
      iy = iy - (iy/2796203) * 2796203
*
*     NB: The dstorf() calls here are CRITICAL: they prevent compilers
*     from replacing the divide by a multiply with the reciprocal of the
*     constant, which results in a completely different sequence of
*     pseudo-random numbers, sigh...  At least Lahey/Fujitsu lf95 and
*     Portland Group, Inc. pgf77, pgf90, pghpf, have this misfeature.
*
*     In the last statement, the outermost dstorf() call also prevents
*     use of a floating-point multiply-add instruction, which has
*     different precision: several RISC architectures have such
*     instructions (Hal/Fujitsu SPARC64-GP, HP PA-RISC, IBM Power and
*     PowerPC, IBM S/390 G5 IEEE 754, Intel IA-64, National
*     Semiconductor NS32381, SGI MIPS IV, ...)
*
      dran = dfloat(iy) / dstorf(2796203.0d+00)
*
      iy = iy * 125
      iy = iy - (iy/2796203) * 2796203
      dran = dran + dstorf(dstorf(dfloat(iy) / dstorf(2796203.0d+00)) /
     X    dstorf(536870912.0d+00))
*     ---------- last card of dran ----------
      END
